libraries/freertos_plus/aws/ota/src/aws_iot_ota_agent.c

```diff
@@ -1796,9 +1796,65 @@ static DocParseErr_t prvInitDocModel( JSON_DocModel_t * pxDocModel,
     return eErr;
 }
 
+/*
+ * Validate the version of the update received.
+ */
+static OTA_Err_t prvValidateUpdateVersion( OTA_FileContext_t * C )
+{
+    DEFINE_OTA_METHOD_NAME( "prvValidateUpdateVersion" );
+
+    OTA_Err_t xErr = kOTA_Err_Uninitialized;
+
+    /* Only check for versions if the target is self */
+    if( xOTA_Agent.ulServerFileID == 0 )
+    {
+        /* Check if update version received is newer than current version.*/
+        if( C->ulUpdaterVersion < xAppFirmwareVersion.u.ulVersion32 )
+        {
+            OTA_LOG_L1( "[%s] The update version is newer than the version on device.\r\n", OTA_METHOD_NAME );
+
+            xErr = kOTA_Err_None;
+        }
+        /* Check if update version received is older than current version.*/
+        else if( C->ulUpdaterVersion > xAppFirmwareVersion.u.ulVersion32 )
+        {
+            OTA_LOG_L1( "[%s] The update version is older than the version on device.\r\n", OTA_METHOD_NAME );
+
+            xErr = kOTA_Err_DowngradeNotAllowed;
+        }
+        /* Check if version reported is the same as the running version. */
+        else if( C->ulUpdaterVersion == xAppFirmwareVersion.u.ulVersion32 )
+        {
+            if( ( xOTA_Agent.pcClientTokenFromJob == NULL ) ||
+                ( strtoul( ( const char * ) xOTA_Agent.pcClientTokenFromJob, NULL, 0 ) == 0U ) )
+            /*lint !e9007 We don't provide a modifiable variable to strtoul.*/
+            {
+                /* The version is the same so either we're not actually the new firmware or
+                 * someone messed up and sent firmware with the same version. In either case,
+                 * this is a failure of the OTA update so reject the job.
+                 */
+                OTA_LOG_L1( "[%s] We rebooted and the version is still the same.\r\n", OTA_METHOD_NAME );
+            }
+            else
+            {
+                OTA_LOG_L1( "[%s] Invalid client token. Device must be rebooted first.\r\n", OTA_METHOD_NAME );
+            }
```

Comment:
This doesn't make sense to me.  Why is the validity of the client token checked after checking that the OTA version  is the same as the running version?  Also, strtoul returns 0 when it can't find a number, not just when the number is zero.

Reply:
Removed redundant check in https://github.com/aws/amazon-freertos/commit/f8bd5081136e1c5e72f9f55f7f185cb52da305ee